proposals/ytt/004-schema-validation/README.md

```diff
@@ -0,0 +1,1523 @@
+
+#### Use Case: Required input
+
+When the author cannot/will not provide a default value for a data value and requires the consumer to provide it.
+
+Where possible, authors are encouraged to use a type-specific "non-empty" validation. As a fallback, they may opt for nullable+not_null.
+
+##### Required input for types with length
+
+**String Values**
+
+The recommended way for an author to indicate required string input is to expect a non-empty string:
+
+```yaml
+#@data/values-schema
+---
+dex:
+  #@assert/validate min_len=1
+  namespace: ""
+  #@assert/validate min_len=1
+  username: ""
+```
+
+This case is so common (it is our understanding that most string-based data values are meant to be non-empty), we offer this convenience:
+
+```yaml
+#@data/values-schema
+#@assert/validate-string-defaults min_len=1
+---
+dex:
+  namespace: ""
+  username: ""
+```
+
+Which is equivalent to the prior example, but applies the default to any descendent that contains a string-typed value. If a particular node already has an `@assert/validate` annotation, it is skipped.
+
+**Array Values**
+
+Similarly, a non-empty value of an array has at least one element:
+
+```yaml
+#@data/values-schema
+---
+dex:
+  oauth2:
+    #@assert/validate min_len=1
+    responseTypes:
+    - ""
+```
+
+Array-typed values are less common. And when present are not necessarily _as_ common to require a non-empty value as string-typed values. So, there is no corresponding shorthand annotation for requiring a non-empty array.
+
+
+##### Required input for types with empty value
+
+For some integer-typed (or float-typed) data values, a zero value may be invalid and thus can represent "empty".
+
+```yaml
+#@data/values-schema
+dex:
+  #@assert/validate min=1024, max=65535
+  port: 0
+```
+
+Authors are encouraged to favor this approach over [marking the data values as nullable](#required-input-for-types-with-no-empty-value) to keep their schema as simple as possible.
+ 
+##### Required input for types with no empty value
+
+Some types have no natural (or easily detectable) "empty" value. In these cases, authors can force the consumer to supply a value by marking the value as "nullable" and constraint it to be "not null":
+
+```yaml
+#@data/values-schema
+---
+#@schema/nullable
+#@assert/validate not_null=("Cloud credentials are required.", True)
+credential:
+  name:
+  secretContents:
+    cloud:
+```
+
+_(also considered: [`@schema/required`](#schemarequired))._
+
+#### Use Case: Enumeration
+
+Values can be constrained to be one of a finite site (i.e. of an enumeration).
+
+```yaml
+#@data/values-schema
+---
+volumeSnapshotLocation:
+  spec:
+    #@assert/validate one_of=["aws", "azure", "vsphere"]
+    provider: ""
+```
+
+#### Use Case: Union Structures
+
+An author can require that exactly one of a set of sibling node has a value:
+
+```yaml
+#@data/values-schema
+---
+dex:
+  #@assert/validate one_not_null=["oidc", "ldap"]
```

Comment:
I found myself wondering what if it was an expression matcher over very specific matchers? i.e make it into a boolean expression instead, like:

`#@assert/validate not_null=("oidc" or ("ldap" and "server_info"))`

This feels like it's much more complex, but also much more flexible (maybe a bad thing)

Reply:
This proposal does allow for this, it's just described further down in the Spec section. A lambda can be used to achieve  what I was asking